### PR TITLE
feat(rulesets): check uniqueness of AsyncAPI messages

### LINKED ISSUE #2223: Add rule to check uniqueness of `messageId`
**User story.**
As a user, I want to know if the messages defined in my AsyncAPI document have a unique `messageId`.

**Describe the solution you'd like**
The following AsyncAPI document is invalid and Spectral should throw errors due to non-uniqueness of `messageId`.

```yaml
asyncapi: '2.4.0'
...

channels:
  someChannel:
    publish:
      message:
        messageId: someId
        ...
    subscribe:
      message:
        messageId: someId
```

**Additional context**
Part of #2100

### PR DESCRIPTION
**Pull request by magicmatatjahu**
Fixes #2223 

**Checklist**

- [X] Tests added / updated
- [X] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

> If indicated yes above, please describe the breaking change(s).
>
> **Remove this quote before creating the PR.**

Part of #2100 

Add rule to check uniqueness `messageId` of AsyncAPI messages.

**Bad Example**

```yaml
channels:
  smartylighting.streetlights.1.0.action.{streetlightId}.turn.on:
    publish:
      message:
        messageId: turnMessage
  smartylighting.streetlights.1.0.action.{streetlightId}.turn.off:
    publish:
      message:
        messageId: turnMessage
```

**Good Example**

```yaml
channels:
  smartylighting.streetlights.1.0.action.{streetlightId}.turn.on:
    publish:
      message:
        messageId: turnOnMessage
  smartylighting.streetlights.1.0.action.{streetlightId}.turn.off:
    publish:
      message:
        messageId: turnOffMessage
```

---

Also fix rule for `operationId` uniqueness - check `operationId` in operation's traits. 

cc @smoya @jonaslagoni

### CODE REVIEW

File: docs/reference/asyncapi-rules.md [hunk 1]
```diff
@@ -219,6 +219,40 @@ components:
 
 **Recommended:** Yes
 
+### asyncapi-message-messageId-uniqueness
+
+`messageId` must be unique across all the messages (except these one defined in the components).
```
Comment (jonaslagoni) [inline docs/reference/asyncapi-rules.md]: Why not all messages? 🤔

File: packages/rulesets/src/asyncapi/functions/utils/getAllMessages.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import type { JsonPath } from '@stoplight/types';
+
+type MessageObject = Record<string, unknown>;
+type AsyncAPI = {
+  channels?: Record<string, { subscribe?: Record<string, unknown>; publish?: Record<string, unknown> }>;
+};
+type Result = { path: JsonPath; message: MessageObject };
+
+export function* getAllMessages(asyncapi: AsyncAPI): IterableIterator<Result> {
+  for (const { path, operation } of getAllOperations(asyncapi)) {
```
Comment (jonaslagoni) [inline packages/rulesets/src/asyncapi/functions/utils/getAllMessages.ts:14]: Dunno why, but I thought we had access to a parsed instance and we could use https://github.com/asyncapi/parser-js/blob/58d0505e23fb4b65333dfd25d6b89b0924622367/lib/models/asyncapi.js#L170

But I guess not 😄

Comment (magicmatatjahu) [reply packages/rulesets/src/asyncapi/functions/utils/getAllMessages.ts:14]: We don't have it that good 😄

File: docs/reference/asyncapi-rules.md (hunk 1 above)
Comment (magicmatatjahu) [reply docs/reference/asyncapi-rules.md]: I've been thinking about it and the components can also be used in other specs and additionally some message do not have to be used at all inside the spec in which they are located, so I check only those that are "actively" involved in the spec, in described operations.

Comment (jonaslagoni) [reply docs/reference/asyncapi-rules.md]: That makes sense yea 🤔 

As long as those messages that are being referenced from the active operations are checked, but they are so all good 👍

**Review by jonaslagoni — DISMISSED** [review]
LGTM 👍

**Review by smoya — DISMISSED** [review]
LGTM!

File: docs/reference/asyncapi-rules.md (hunk 1 above)
Comment (pamgoodrich) [inline docs/reference/asyncapi-rules.md]: "...except these one defined..." > "except those defined..."

**Review by pamgoodrich — DISMISSED** [review]
Approve the doc file with one small change.

File: docs/reference/asyncapi-rules.md (hunk 1 above)
Comment (P0lip) [reply docs/reference/asyncapi-rules.md]:
```suggestion
`messageId` must be unique across all the messages (except those one defined in the components).
```

**Review by P0lip — APPROVED** [review]
Nice work

**Comment by P0lip** [comment]
Apologies for a delay - was taking time off, and this one felt through my cracks.
Do feel free to merge next time :)

**Comment by magicmatatjahu** [comment]
@P0lip @pamgoodrich Thanks for review and suggestion! Just today I saw this review but I also saw that it is merged 😄  Nothing is wrong, we help in this project so we can wait :)

**Pull request merged**